README.md
# ArduinoFortune

Display a fortune on an LCD.

Tapping the screen will change the fortune.

## Hardware needed:
+ Arduino Uno
+ 2.8 in TFT LCD

Both are available from [Banggood](https://www.banggood.com/Geekcreit-UNO-R3-Improved-Version-2_8TFT-LCD-Touch-Screen-2_4TFT-Touch-Screen-Display-Module-Kit-p-1428291.html) for cheap.

## Overview

The splitFortunes.py program takes the Linux Fortune files and splits each fortune into a single file, grouping files in subdirectories of 255 files each.  This is done to increase the speed of the fortune displaying.

I tried reading the Linux fortune files, but if a large fortune file was chosen and a fortune near the end of that file was chosen, it would take upwards of 8 seconds for the fortune to display.  A co-worker suggested the single-fortune-in-a-file solution which greatly improved things, but the SD card library goes through the directory sequentially looking for the file to open.  So if a large numbered fortune was chosen, it would again take upwards of 8 seconds to display.  The solution was to group the fortunes into subdirectories.  Now they display very quickly.

The splitFortunes.py also reformats the fortunes to fit on the display and will quietly drop fortunes that have too many lines to fit on the screen.

All that makes the Arduino code much simpler.  It simply randomly chooses a fortune file.  A simple division tells it what subdirectory it's in.  It then reads the file and writes it to the display.

## Software

+ sdCard/ - the files that need to go on the micro SD card in the LCD shield.
+ sdCard/splitFortunes.py - Reads the Linux Fortune files and splits each fortune into an individual file, creating the subdirectories.
+ sdCard/fortune.py - The Fortune module that I use to read Linux Fortune files.
+ sdCard/<everything else> - The Linux fortune files.  These came from the repo for the Linux Fortune program.  I did do some work to eliminate some fortunes (mostly in the ascii-art) that simply won't format well.
+ Fortune/Fortune.ino - The Arduino script to display the fortunes.


sdCard/news
A help wanted add for a photo journalist asked the rhetorical question:

If you found yourself in a situation where you could either save a drowning man, or you could take a Pulitzer prize winning photograph of him drowning, what shutter speed and setting would you use?
    -- Paul Harvey
%
A Hen Brooding Kittens
    A friend informs us that he saw at the Novato ranch, Marin county, a few days since, a hen actually brooding and otherwise caring for three kittens!  The gentleman upon whose premises this strange event is transpiring says the hen adopted the kittens when they were but a few days old, and that she has devoted them her undivided care for several weeks past.  The young felines are now of respectable size, but they nevertheless follow the hen at her cluckings, and are regularly brooded at night beneath her wings.
    -- Sacramento Daily Union, July 2, 1861
%
A journalist, thrilled over his dinner, asked the chef for the recipe.  Retorted the chef, "Sorry, we have the same policy as you journalists, we never reveal our sauce."
%
A Mexican newspaper reports that bored Royal Air Force pilots stationed on the Falkland Islands have devised what they consider a marvelous new game.  Noting that the local penguins are fascinated by airplanes, the pilots search out a beach where the birds are gathered and fly slowly along it at the water's edge.  Perhaps ten thousand penguins turn their heads in unison watching the planes go by, and when the pilots turn around and fly back, the birds turn their heads in the opposite direction, like spectators at a slow-motion tennis match.  Then, the paper reports "The pilots fly out to sea and directly to the penguin colony and overfly it.  Heads go up, up, up, and ten thousand penguins fall over gently onto their backs.
    -- Audobon Society Magazine
%
A New Way of Taking Pills
    A physician one night in Wisconsin being disturbed by a burglar, and having no ball or shot for his pistol, noiselessly loaded the weapon with small, hard pills, and gave the intruder a "prescription" which he thinks will go far towards curing the rascal of a very bad ailment.
    -- Nevada Morning Transcript, January 30, 1861
%
A newspaper is a circulating library with high blood pressure.
    -- Arthure "Bugs" Baer
%
A prominent broadcaster, on a big-game safari in Africa, was taken to a watering hole where the life of the jungle could be observed. As he looked down from his tree platform and described the scene into his tape recorder, he saw two gnus grazing peacefully. So preoccupied were they that they failed to observe the approach of a pride of lions led by two magnificent specimens, obviously the leaders. The lions charged, killed the gnus, and dragged them into the bushes where their feasting could not be seen.  A little while later the two kings of the jungle emerged and the radioman recorded on his tape: "Well, that's the end of the gnus and here, once again, are the head lions."
%
"A raccoon tangled with a 23,000 volt line today.  The results blacked out 1400 homes and, of course, one raccoon."
    -- Steel City News
%
A young girl once committed suicide because her mother refused her a new bonnet.  Coroner's verdict: "Death from excessive spunk."
    -- Sacramento Daily Union, September 13, 1860
%
Advertisements contain the only truths to be relied on in a newspaper.
    -- Thomas Jefferson
%
After two or three weeks of this madness, you begin to feel As One with the man who said, "No news is good news." In twenty-eight papers, only the rarest kind of luck will turn up more than two or three articles of any interest...  but even then the interest items are usually buried deep around paragraph 16 on the jump (or "Cont.  on ...") page...

The Post will have a story about Muskie making a speech in Iowa.  The Star will say the same thing, and the Journal will say nothing at all.  But the Times might have enough room on the jump page to include a line or so that says something like: "When he finished his speech, Muskie burst into tears and seized his campaign manager by the side of the neck.  They grappled briefly, but the struggle was kicked apart by an oriental woman who seemed to be in control."

Now that's good journalism.  Totally objective; very active and straight to the point.
    -- Hunter S. Thompson, "Fear and Loathing '72"
%
All newspaper editorial writers ever do is come down from the hills after the battle is over and shoot the wounded.
%
An editor is one who separates the wheat from the chaff and prints the chaff.
    -- Adlai Stevenson
%
"... And remember: if you don't like the news, go out and make some of
your own."
    -- "Scoop" Nisker, KFOG radio reporter Preposterous Words
%
And that's the way it is...
    -- Walter Cronkite
%
Earth Destroyed by Solar Flare -- film clips at eleven.
%
Every journalist has a novel in him, which is an excellent place for it.
%
Everything you read in newspapers is absolutely true, except for that rare story of which you happen to have first-hand knowledge.
    -- Erwin Knoll
%
FLASH!
Intelligence of mankind decreasing.
Details at ... uh, when the little hand is on the ....
%
... Had this been an actual emergency, we would have fled in terror, and you would not have been informed.
%
I only know what I read in the papers.
    -- Will Rogers
%
I read the newspaper avidly.  It is my one form of continuous fiction.
    -- Aneurin Bevan
%
I really look with commiseration over the great body of my fellow citizens who, reading newspapers, live and die in the belief that they have known something of what has been passing in their time.
    -- H. Truman
%
If I were to walk on water, the press would say I'm only doing it because I can't swim.
    -- Bob Stanfield
%
If you lose your temper at a newspaper columnist, he'll get rich, or famous or both.
%
In a medium in which a News Piece takes a minute and an "In-Depth" Piece takes two minutes, the Simple will drive out the Complex.
    -- Frank Mankiewicz
%

sdCard/fortune.py
#!/usr/bin/env python3

import os.path

class Fortune:

    files=["art", "ascii-art", "computers", "cookie", "debian", 
           "definitions", "disclaimer", "education", "ethnic", 
           "food", "fortunes", "goedel", "humorists", "kids", 
           "law", "linux", "literature", "love", "magic", 
           "medicine", "men-women", "miscellaneous", "news",
           "paradoxum", "people", "pets", "platitudes", "politics",
           "riddles", "science", "sports", "startrek",
           "translate-me", "wisdom", "work", "zippy"]
    counts={}

    def get_counts_from_cache(self,cache_file):
        with open(cache_file) as file:
            for line in file:
                parts=line.strip().split(" ")
                self.counts[parts[0]] = int(parts[1])


    def create_count_cache(self,cache_file):
        with open(cache_file,"w") as file:
            for name in self.files:
                count=self.get_fortune_count(name)
                file.write("{} {}\n".format(name,count))
                self.counts[name] = count


    def cache_fortune_counts(self):
        cache_file="count_cache"

        if os.path.exists(cache_file):
            self.get_counts_from_cache(cache_file)
        else:
            self.create_count_cache(cache_file)


    def get_fortune_count(self,fortune_file):
        fortune_count=0

        with open(fortune_file) as file:
            for line in file:
                if line[0:1] == "%":
                    fortune_count = fortune_count +1

        return(fortune_count)


    def get_fortune_text(self,fortune_file, fortune_num):
        fortune_count=0
        fortune=[]

        with open(fortune_file) as file:
            for line in file:
                if line[0:1] == "%":
                    fortune_count = fortune_count +1
                    if len(fortune) > 0:
                        return fortune

                if fortune_count == fortune_num:
                    if line[0:1] != "%":
                        fortune.append(line.rstrip())

        return("Error finding fortune")


    def get_fortune(self, fortune_file,fortune_num):
        return(self.get_fortune_text(fortune_file,fortune_num))


sdCard/splitFortunes.py
#!/usr/bin/env python3

# This program takes all the fortunes and puts a single fortune into a single file.
# It will also reformat the fortune to fit on the LCD, dropping any fortune that's too long for the LCD (See max_length and max_lines).

# The time required for the Arduino to get a certain fortune is proportional to its location in the directory.
# So fortune 00000.txt will come up fast, but fortune 12500.txt will take 10 seconds.
# So we put the fortunes into subdirectories of no more than 255 files.

import textwrap
import os
from fortune import Fortune

files = [
    "art", "ascii-art", "computers", "cookie", "debian", "definitions",
    "disclaimer", "education", "ethnic", "food", "fortunes", "goedel",
    "humorists", "kids", "law", "linux", "literature", "love", "magic",
    "medicine", "men-women", "miscellaneous", "news", "paradoxum", "people",
    "pets", "platitudes", "politics", "riddles", "science", "sports",
    "startrek", "translate-me", "wisdom", "work", "zippy"
]


def format_fortune(fortune, max_length):
    formatted = []

    for line in fortune:
        wrapped_text = textwrap.wrap(line, max_length)
        for new_line in wrapped_text:
            formatted.append(new_line)

    return (formatted)

# Number of chars per line
max_length = 50
# Number of lines on the screen
max_lines = 30

if __name__ == "__main__":
    f = Fortune()

    # file number
    fn = 0

    # loop over all the fortune files
    for filename in files:
        print("Processing {0}".format(filename))
        num_fortunes = f.get_fortune_count(filename)

        # loop over each fortune in the fortune file
        for i in range(num_fortunes):
            directory=int(fn/255)

            directory_name="f{0:03}".format(directory)
            if not os.path.exists(directory_name):
               os.mkdir(directory_name)

            output_file_name = "{0}/{1:05}.txt".format(directory_name,fn)

            # get that fortune
            fortune_lines = format_fortune(f.get_fortune(filename, i), max_length)

            # If the formatted fortune is "too big", bypass it
            if (len(fortune_lines) <= max_lines):
                # Open the new fortune file
                with open(output_file_name, "w") as new_file:
                    print(output_file_name)
                    # Calculate the number of lines to sort of center it on on the screen
                    padding_lines = int((max_lines - len(fortune_lines)) / 2)
                    # If there's at least 1 line
                    if padding_lines > 0:
                        # write out the padding line
                        for x in range(padding_lines):
                            new_file.write("\n")

                    # now write out the formatted fortune lines
                    for line in fortune_lines:
                        new_file.write("{0}\n".format(line))

                fn = fn + 1

